File: FE-Customer/src/api/productService.ts
import axiosClient, { ApiResponse } from './axiosClient';
import { API_ENDPOINTS } from './constants';

export interface Product {
  id: number;
  productCode: string;
  name: string;
  slug: string;
  shortDescription?: string;
  description?: string;
  imageUrl?: string;
  salePrice: number;
  costPrice?: number;
  stockQuantity: number;
  soldQuantity: number;
  status: boolean;
  isFeatured: boolean;
  categoryId: number;
  categoryName?: string;
  createdAt?: string;
}

export interface ProductsResponse {
  items: Product[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages?: number;
  hasNextPage?: boolean;
  hasPreviousPage?: boolean;
}

export type ProductSortOption =
  | 'newest'
  | 'price-asc'
  | 'price-desc'
  | 'best-seller'
  | 'name-asc'
  | 'name-desc';

export interface ProductListParams {
  page?: number;
  pageSize?: number;
  search?: string;
  categoryId?: number;
  category?: string;
  status?: boolean;
  minPrice?: number;
  maxPrice?: number;
  sort?: ProductSortOption | string;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
}

const mapSortParams = (sort?: ProductListParams['sort']) => {
  switch (sort) {
    case 'price-asc':
      return { sortBy: 'price', sortDir: 'asc' as const };
    case 'price-desc':
      return { sortBy: 'price', sortDir: 'desc' as const };
    case 'best-seller':
      return { sortBy: 'sold', sortDir: 'desc' as const };
    case 'name-asc':
      return { sortBy: 'name', sortDir: 'asc' as const };
    case 'name-desc':
      return { sortBy: 'name', sortDir: 'desc' as const };
    case 'newest':
    default:
      return { sortBy: 'created_at', sortDir: 'desc' as const };
  }
};

const buildProductQueryParams = (params: ProductListParams = {}) => {
  const {
    search,
    categoryId,
    status = true,
    minPrice,
    maxPrice,
    sort,
    sortBy,
    sortDir,
    category: _category,
    ...rest
  } = params;

  const query: Record<string, string | number | boolean> = {
    ...rest,
    status,
  };

  if (search?.trim()) {
    query.name = search.trim();
  }

  if (typeof categoryId === 'number') {
    query.categoryId = categoryId;
  }

  if (typeof minPrice === 'number') {
    query.minPrice = minPrice;
  }

  if (typeof maxPrice === 'number') {
    query.maxPrice = maxPrice;
  }

  const resolvedSort = sortBy
    ? { sortBy, sortDir: sortDir ?? 'desc' }
    : mapSortParams(sort);

  query.sortBy = resolvedSort.sortBy;
  query.sortDir = resolvedSort.sortDir;

  return query;
};

export const productService = {
  getAll: async (
    params?: ProductListParams
  ): Promise<ApiResponse<ProductsResponse>> => {
    return axiosClient.get(API_ENDPOINTS.PRODUCTS, {
      params: buildProductQueryParams(params),
    });
  },

  getAllPages: async (params?: ProductListParams): Promise<Product[]> => {
    const pageSize = Math.max(params?.pageSize ?? 100, 1);
    const items: Product[] = [];
    let page = 1;
    let totalPages = 1;

    do {
      const response = await productService.getAll({
        ...params,
        page,
        pageSize,
      });

      if (!response.success || !response.data) {
        throw new Error(response.error || 'Khong the tai danh sach san pham');
      }

      items.push(...response.data.items);

      totalPages = response.data.totalPages
        ?? Math.max(1, Math.ceil(response.data.totalCount / response.data.pageSize));

      if (response.data.items.length === 0) {
        break;
      }

      page += 1;
    } while (page <= totalPages);

    return items;
  },

  getById: (id: string | number): Promise<ApiResponse<Product>> => {
    return axiosClient.get(API_ENDPOINTS.PRODUCT_DETAIL(id));
  },

  getBySlug: (slug: string): Promise<ApiResponse<Product>> => {
    return axiosClient.get(API_ENDPOINTS.PRODUCT_BY_SLUG(slug));
  },

  search: (
    keyword: string,
    page = 1,
    pageSize = 12
  ): Promise<ApiResponse<ProductsResponse>> => {
    return productService.getAll({ page, pageSize, search: keyword });
  },

  getFeatured: (top = 12): Promise<ApiResponse<Product[]>> => {
    return axiosClient.get(API_ENDPOINTS.PRODUCT_FEATURED, { params: { top } });
  },
};

File: API/Services/API.Products/Services/ProductService.cs
using API.Products.Constants;
using API.Products.Interfaces;
using API.Products.Models.Commands;
using API.Products.Models.DTOs;
using API.Products.Models.Queries;
using Microsoft.EntityFrameworkCore;
using netcore.Commons.Exceptions;
using netcore.Commons.Interfaces;
using netcore.Commons.Models;
using netcore.Entities.Entities;
using netcore.Entities.Interfaces;

namespace API.Products.Services;

public class ProductService : IProductService
{
    private readonly IUnitOfWork _uow;
    private readonly IObjectStorageService _storage;
    private readonly ILogger<ProductService> _logger;

    public ProductService(IUnitOfWork uow, IObjectStorageService storage, ILogger<ProductService> logger)
    {
        _uow = uow;
        _storage = storage;
        _logger = logger;
    }

    public async Task<PagedResult<ProductListDto>> GetAllAsync(ProductQuery query, CancellationToken ct = default)
    {
        var q = _uow.Products.Query()
            .Include(p => p.Category)
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(query.Name))
            q = q.Where(p => p.Name.Contains(query.Name));

        if (query.CategoryId.HasValue)
            q = q.Where(p => p.CategoryId == (decimal)query.CategoryId.Value);

        if (query.Status.HasValue)
            q = q.Where(p => p.Status == query.Status.Value);

        if (query.IsFeatured.HasValue)
            q = q.Where(p => p.IsFeatured == query.IsFeatured.Value);

        if (query.MinPrice.HasValue)
            q = q.Where(p => p.SalePrice >= query.MinPrice.Value);

        if (query.MaxPrice.HasValue)
            q = q.Where(p => p.SalePrice <= query.MaxPrice.Value);

        q = (query.SortBy?.ToLower(), query.SortDir?.ToLower()) switch
        {
            ("price", "asc") => q.OrderBy(p => p.SalePrice),
            ("price", _) => q.OrderByDescending(p => p.SalePrice),
            ("name", "asc") => q.OrderBy(p => p.Name),
            ("name", _) => q.OrderByDescending(p => p.Name),
            ("sold", _) => q.OrderByDescending(p => p.SoldQuantity),
            _ => q.OrderByDescending(p => p.CreatedAt)
        };

        var totalCount = await q.CountAsync(ct);
        var pageSize = Math.Min(query.PageSize, ProductConstants.MaxPageSize);
        var page = Math.Max(query.Page, 1);

        var items = await q
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(p => MapToListDto(p))
            .ToListAsync(ct);

        return new PagedResult<ProductListDto>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<ProductDto> GetByIdAsync(long id, CancellationToken ct = default)
    {
        var product = await _uow.Products.Query()
            .Include(p => p.Category)
            .Include(p => p.ProductImages)
            .FirstOrDefaultAsync(p => p.Id == (decimal)id, ct)
            ?? throw new NotFoundException("Sản phẩm", id);

        return MapToDto(product);
    }

    public async Task<ProductDto> GetBySlugAsync(string slug, CancellationToken ct = default)
    {
        var product = await _uow.Products.Query()
            .Include(p => p.Category)
            .Include(p => p.ProductImages)
            .FirstOrDefaultAsync(p => p.Slug == slug, ct)
            ?? throw new NotFoundException($"Không tìm thấy sản phẩm với slug: {slug}");

        return MapToDto(product);
    }

    public async Task<List<ProductListDto>> GetNoiBatAsync(int top = 8, CancellationToken ct = default)
    {
        return await _uow.Products.Query()
            .Include(p => p.Category)
            .Where(p => p.IsFeatured == true && p.Status == true)
            .OrderByDescending(p => p.SoldQuantity)
            .Take(top)
            .Select(p => MapToListDto(p))
            .ToListAsync(ct);
    }

    public async Task<ProductDto> UpsertAsync(UpsertProductRequest request, CancellationToken ct = default)
    {
        _ = await _uow.Categories.FirstOrDefaultAsync(c => c.Id == (decimal)request.CategoryId, ct)
            ?? throw new NotFoundException("Danh mục", request.CategoryId);

        Product product;

        if (request.Id.HasValue)
        {
            product = await _uow.Products.Query()
                .Include(p => p.ProductImages)
                .FirstOrDefaultAsync(p => p.Id == (decimal)request.Id.Value, ct)
                ?? throw new NotFoundException("Sản phẩm", request.Id.Value);

            product.CategoryId = (decimal)request.CategoryId;
            product.Name = request.Name;
            product.Slug = string.IsNullOrWhiteSpace(request.Slug) ? GenerateSlug(request.Name) : request.Slug;
            product.Sku = request.Sku;
            product.ShortDescription = request.ShortDescription;
            product.Description = request.Description;
            product.CostPrice = request.CostPrice;
            product.SalePrice = request.SalePrice;
            product.StockQuantity = (decimal)request.StockQuantity;
            product.WeightGrams = request.WeightGrams;
            product.Status = request.Status;
            product.IsFeatured = request.IsFeatured;
            product.UpdatedAt = DateTime.UtcNow;

            _uow.Products.Update(product);
        }
        else
        {
            product = new Product
            {
                CategoryId = (decimal)request.CategoryId,
                ProductCode = $"SP{DateTime.UtcNow:yyyyMMddHHmmss}",
                Name = request.Name,
                Slug = string.IsNullOrWhiteSpace(request.Slug) ? GenerateSlug(request.Name) : request.Slug,
                Sku = request.Sku,
                ShortDescription = request.ShortDescription,
                Description = request.Description,
                CostPrice = request.CostPrice,
                SalePrice = request.SalePrice,
                StockQuantity = (decimal)request.StockQuantity,
                SoldQuantity = 0,
                WeightGrams = request.WeightGrams,
                Status = request.Status,
                IsFeatured = request.IsFeatured,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            await _uow.Products.AddAsync(product, ct);
        }

        var productFolder = $"products/{product.ProductCode}";
        product.ImageUrl = await _storage.ReplaceImageAsync(
            request.ImageFile,
            product.ImageUrl,
            $"{productFolder}/primary",
            request.RemoveImage,
            request.ImageUrl,
            ct);

        await _uow.SaveChangesAsync(ct);
        await SyncProductImagesAsync(product, request.Images, productFolder, ct);
        await _uow.SaveChangesAsync(ct);

        _logger.LogInformation("Product upserted: {ProductCode} - {Name}", product.ProductCode, product.Name);
        return await GetByIdAsync((long)product.Id, ct);
    }

    public async Task DeleteAsync(DeleteProductRequest request, CancellationToken ct = default)
    {
        var product = await _uow.Products.Query()
            .Include(p => p.ProductImages)
            .FirstOrDefaultAsync(p => p.Id == (decimal)request.Id, ct)
            ?? throw new NotFoundException("Sản phẩm", request.Id);

        foreach (var image in product.ProductImages)
        {
            await _storage.RemoveIfManagedAsync(image.ImageUrl, ct);
        }

        await _storage.RemoveIfManagedAsync(product.ImageUrl, ct);
        _uow.ProductImages.RemoveRange(product.ProductImages);
        _uow.Products.Remove(product);
        await _uow.SaveChangesAsync(ct);

        _logger.LogInformation("Product deleted: {Id}", request.Id);
    }

    private async Task SyncProductImagesAsync(
        Product product,
        List<UpsertProductImageRequest> requests,
        string productFolder,
        CancellationToken ct)
    {
        if (requests.Count == 0)
        {
            return;
        }

        var existingImages = await _uow.ProductImages.Query()
            .Where(i => i.ProductId == product.Id)
            .ToListAsync(ct);

        foreach (var item in requests)
        {
            if (item.Id.HasValue)
            {
                var existing = existingImages.FirstOrDefault(i => i.Id == (decimal)item.Id.Value);
                if (existing is null)
                {
                    continue;
                }

                if (item.Remove)
                {
                    await _storage.RemoveIfManagedAsync(existing.ImageUrl, ct);
                    _uow.ProductImages.Remove(existing);
                    continue;
                }

                existing.ImageUrl = await _storage.ReplaceImageAsync(
                    item.ImageFile,
                    existing.ImageUrl,
                    $"{productFolder}/gallery",
                    false,
                    item.ImageUrl,
                    ct) ?? existing.ImageUrl;
                existing.AltText = item.AltText;
                existing.IsPrimary = item.IsPrimary;
                existing.DisplayOrder = item.DisplayOrder;
                _uow.ProductImages.Update(existing);
                continue;
            }

            if (item.Remove || (item.ImageFile is null && string.IsNullOrWhiteSpace(item.ImageUrl)))
            {
                continue;
            }

            var imageUrl = await _storage.ReplaceImageAsync(
                item.ImageFile,
                null,
                $"{productFolder}/gallery",
                false,
                item.ImageUrl,
                ct);

            if (string.IsNullOrWhiteSpace(imageUrl))
            {
                continue;
            }

            await _uow.ProductImages.AddAsync(new ProductImage
            {
                ProductId = product.Id,
                ImageUrl = imageUrl,
                AltText = item.AltText,
                IsPrimary = item.IsPrimary,
                DisplayOrder = item.DisplayOrder,
                CreatedAt = DateTime.UtcNow
            }, ct);
        }
    }

    private static string GenerateSlug(string name)
    {
        return name.ToLowerInvariant()
            .Replace(" ", "-")
            .Replace("đ", "d")
            .Trim('-');
    }

    private static ProductListDto MapToListDto(Product p) => new()
    {
        Id = (long)p.Id,
        ProductCode = p.ProductCode,
        Name = p.Name,
        Slug = p.Slug,
        ShortDescription = p.ShortDescription,
        Description = p.Description,
        ImageUrl = p.ImageUrl,
        SalePrice = p.SalePrice,
        StockQuantity = (int)p.StockQuantity,
        SoldQuantity = (int)p.SoldQuantity,
        Status = p.Status ?? true,
        IsFeatured = p.IsFeatured ?? false,
        CategoryId = (long)p.CategoryId,
        CategoryName = p.Category?.Name,
        CreatedAt = p.CreatedAt
    };

    private static ProductDto MapToDto(Product p) => new()
    {
        Id = (long)p.Id,
        CategoryId = (long)p.CategoryId,
        CategoryName = p.Category?.Name,
        ProductCode = p.ProductCode,
        Sku = p.Sku,
        Name = p.Name,
        Slug = p.Slug,
        ShortDescription = p.ShortDescription,
        Description = p.Description,
        ImageUrl = p.ImageUrl,
        CostPrice = p.CostPrice,
        SalePrice = p.SalePrice,
        StockQuantity = (int)p.StockQuantity,
        SoldQuantity = (int)p.SoldQuantity,
        WeightGrams = p.WeightGrams,
        Status = p.Status ?? true,
        IsFeatured = p.IsFeatured ?? false,
        CreatedAt = p.CreatedAt,
        UpdatedAt = p.UpdatedAt,
        Images = p.ProductImages?.Select(i => new ProductImageDto
        {
            Id = (long)i.Id,
            ImageUrl = i.ImageUrl,
            AltText = i.AltText,
            IsPrimary = i.IsPrimary ?? false,
            DisplayOrder = (int)i.DisplayOrder
        }).OrderBy(i => i.DisplayOrder).ToList() ?? []
    };
}
